### PR TITLE
Android support

### PR DESCRIPTION
**Pull request by loonycyborg**
Current way to build:
`
cd packaging/android/build-android-dependencies
./build-android-deps.sh
mkdir /tmp/wesnoth && cd /tmp/wesnoth
scons -Y <path/to/checkout> host=android-armeabi-v7a prefix=/tmp/android-prefix/armeabi-v7a
`
Fixes issue #7477

### CODE REVIEW

**Comment by Pentarctagon** [comment]
Is it possible to add a CI job for this? Also, has running it on an Android phone been tested at all?

**Comment by loonycyborg** [comment]
I managed to run it on android phone without a crash but it's not too useful yet since even data isn't loaded, hence draft.

**Comment by loonycyborg** [comment]
> Is it possible to add a CI job for this?

Yes, but need to think how best to get a container/runner with android sdk/ndk installed.

**Comment by gfgtdf** [comment]
Whats the current state of this? Do you have a todo list on what you still have planned to do here?

**Comment by babaissarkar** [comment]
Was doing some testing on this on my local branch and noticed the following issues:
- [ ] Text rendering does not work. Either causes crash or renders boxes, an indication that pango/cairo cannot find the font.
- [ ] Threading crashes. `call_in_main_thread` especially has a tendency to cause random SDL_Thread crashes. The Event handler does not work as well, but I have a patch for that, which I'll shortly post.

**Comment by loonycyborg** [comment]
Most likely it's because of fontconfig. Can you share logs?

**Comment by babaissarkar** [comment]
> Most likely it's because of fontconfig. Can you share logs?

It did turn out to be a fontconfig issue. Fontconfig was searching for fonts.conf in the wrong place.

https://github.com/wesnoth/wesnoth/blob/c5cb0f104fae4da82c54caf85ce3047c9c4a0e97/src/font/font_config.cpp#L122-L130

When this is called `game_config::path` was still not set to the correct value, and defaulted to wesnoth's build directory on my PC, which obviously doesn't exist on my phone.

**Comment by loonycyborg** [comment]
It seems like it couldn't show a dialog. Do you set targeted android api in manifest somewhere? Some minimum api version is required to support dpi-aware apps like wesnoth.

**Comment by babaissarkar** [comment]
> Some minimum api version is required to support dpi-aware apps like wesnoth.
Will check.

**Comment by soliton-** [comment]
IMO the thread ID stuff does not belong here. Doesn't look like something android specific. Should just be committed to master.

**Comment by babaissarkar** [comment]
@loonycyborg can this branch be rebased on current master?

**Comment by babaissarkar** [comment]
Made some progress on [my 1.19 branch](https://github.com/babaissarkar/wesnoth/tree/android-support-1.19). The tutorial can be played and most of the dialogs and the map editor are working. Also wesnoth's existing touchscreen supported works fairly well on Android as well: touch to select hex, double tap to right click, drag to move unit/attack.

https://drive.google.com/file/d/1GqTbdxbAh05svG8MGYOr0U19lh1sGYN8/view?usp=drive_link

(Uploaded to my gdrive then linked here since it's too big as an github attachment.)

**Remaining issues:**
- [x] No sound/music (https://github.com/wesnoth/wesnoth/commit/ef483533fee0ed516eec229d5dddc10da586e39a)
- [x] Fonts not correctly used unless specified via Pango markup (fixed by reverting my debug changes to text.cpp)
- [x] Opening links not working (778ec07608cdf39936589f9b0044d4d9763d1272) (Related: implementation of other `desktop::` methods)
- [x] Data download mechanism (Added but needs refining)

**Minor**

- [x] OS info in About dialog (partially added, shows `Android (API level XX)` on About dialog)
- [x] Battery indication (https://github.com/wesnoth/wesnoth/commit/c2cd5b829b3dd95cc1e67b2f2e6fddae2cd71d2d)
- [x] Everything looks tiny (Font size only so far)
- [ ] Pause/resume causes Wesnoth to restart, even if it's for a very small time. (SDL_HINT_ANDROID_BLOCK_ON_PAUSE? also https://github.com/libsdl-org/SDL/issues/3193 might be relevant.)
- [ ] Tutorial instructions need to be adjusted for touchscreen
- [ ] No hover, so hover-heavy things (tooltips) may need to be redone.
- [ ] No way to change speed/scrolling in Credits
- [ ] Pinch Zoom In/Out in Map
- [x] Testing needed: translations and locale detection
- [ ] (Probably more... but these previous ones are obvious)

**Comment by babaissarkar** [comment]
@loonycyborg https://github.com/wesnoth/wesnoth/pull/7707/commits/c23a9ee3b371451d270c70a5d940a304624c31fa bumps the SDL2 library version without updating the SDL java files. Suggest updating both to 2.30.12 (https://github.com/libsdl-org/SDL/tree/release-2.30.x/android-project/app/src/main/java/org/libsdl/app).

**Comment by loonycyborg** [comment]
I would prefer to merge your #9825 first. Though I hope it won't cause bad things to happen because I used merge instead of rebase to update against master in the past.

**Review by Vultraz — CHANGES_REQUESTED** [review]

File: src/video.cpp [hunk 1]
```diff
@@ -542,7 +556,7 @@ void reset_render_target()
 texture get_render_target()
 {
 	// This should always be up-to-date, but assert for sanity.
-	assert(current_render_target_ == SDL_GetRenderTarget(get_renderer()));
+	// assert(current_render_target_ == SDL_GetRenderTarget(get_renderer()));
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/video.cpp:559]: Assert fails on Android.

File: source_lists/wesnoth [hunk 2]
```diff
@@ -306,7 +306,6 @@ play_controller.cpp
 playmp_controller.cpp
 playsingle_controller.cpp
 playturn_network_adapter.cpp
-preferences/preferences.cpp
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply source_lists/wesnoth:309]: Alright, will readd.

[the remark this answers is not in the document]
Comment (Vultraz) [reply source_lists/wesnoth:309]: I mean, it doesn't need to be if it works without it, it's just an random change to have here. Was wondering if it was needed to make something compile.

File: src/wesnoth.cpp [hunk 3]
```diff
@@ -597,6 +604,7 @@ static int process_command_args(commandline_options& cmdline_opts)
 		return 2;
 	}
 
+
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/wesnoth.cpp]: done

File: src/video.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/video.cpp]: Put it under `#ifndef __ANDROID`

File: source_lists/wesnoth (hunk 2 above)
[the remark this answers is not in the document]
Comment (babaissarkar) [reply source_lists/wesnoth:309]: Looks like `preferences/preferences.cpp` exists in both `wesnoth` and `libwesnoth`, is that intentional?

File: src/desktop/battery_info.cpp [hunk 4]
```diff
@@ -27,6 +27,11 @@
 #include "desktop/dbus_features.hpp"
 #endif
 
+#ifdef __ANDROID__
+#include <SDL2/SDL.h> // For SDL Android functions
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/desktop/battery_info.cpp:31]: updated to use `SDL_system.h`

File: src/wesnoth.cpp [hunk 5]
```diff
@@ -781,7 +786,6 @@ static int do_gameloop(commandline_options& cmdline_opts)
 	gui2::switch_theme(prefs::get().gui2_theme());
 	const gui2::event::manager gui_event_manager;
 
-	// if the log directory is not writable, then this is the error condition so show the error message.
```
Comment (soliton-) [inline src/wesnoth.cpp:784]: What is the point of this?

File: src/wesnoth.cpp [hunk 6]
```diff
@@ -997,6 +1003,13 @@ int main(int argc, char** argv)
 	setenv("PANGOCAIRO_BACKEND", "fontconfig", 0);
 #endif
 
+#ifdef __ANDROID__
+	game_config::path = SDL_AndroidGetExternalStoragePath() + std::string("/gamedata");
+	setenv("PANGOCAIRO_BACKEND", "fontconfig", 1);
+	setenv("FONTCONFIG_PATH", (game_config::path + "/fonts").c_str(), 1);
+	setenv("SDL_HINT_AUDIODRIVER", "android", 1);
```
Comment (soliton-) [inline src/wesnoth.cpp:1010]: What is the benefit of the last argument (overwrite) being 1 in each call?

File: src/wesnoth.cpp [hunk 7]
```diff
@@ -997,6 +1003,13 @@ int main(int argc, char** argv)
 	setenv("PANGOCAIRO_BACKEND", "fontconfig", 0);
 #endif
 
+#ifdef __ANDROID__
+	game_config::path = SDL_AndroidGetExternalStoragePath() + std::string("/gamedata");
```
Comment (soliton-) [inline src/wesnoth.cpp:1007]: This should be done with the game data path setting in `process_command_args`.

File: src/wesnoth.cpp (hunk 6 above)
Comment (babaissarkar) [reply src/wesnoth.cpp:1010]: I thought > 0 values caused overwrite of the corresponding env. var.?

File: src/wesnoth.cpp [hunk 8]
```diff
@@ -408,12 +413,12 @@ static int process_command_args(commandline_options& cmdline_opts)
...
 
 	if(cmdline_opts.usercache_path) {
-		std::cout << filesystem::get_cache_dir();
+		PLAIN_LOG << filesystem::get_cache_dir();
 		return 0;
 	}
 
 	if(cmdline_opts.userdata_path) {
-		std::cout << filesystem::get_user_data_dir();
+		PLAIN_LOG << filesystem::get_user_data_dir();
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/wesnoth.cpp:421]: Can revert that, but I thought `PLAIN_LOG` was the norm? Any particular reason it uses `std::cout` instead?

File: src/wesnoth.cpp (hunk 5 above)
Comment (babaissarkar) [reply src/wesnoth.cpp:784]: most likely a merge accident, will revert.

File: src/wesnoth.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Pentarctagon) [reply src/wesnoth.cpp:421]: The difference is PLAIN_LOG goes to `std::cerr` rather than `std::cout` (when not logging to a file).

File: src/wesnoth.cpp [hunk 9]
```diff
@@ -804,22 +808,24 @@ static int do_gameloop(commandline_options& cmdline_opts)
...
+			return 1;
 		}
 
 		if(!game_config::no_addons && !cmdline_opts.noaddons)  {
 			loading_screen::progress(loading_stage::refresh_addons);
 
 			refresh_addon_version_info_cache();
 		}
+
+		return 0;
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/wesnoth.cpp]: the change proved to be unnecessary, so made it same as master.

File: src/wesnoth.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/wesnoth.cpp]: changed back to `std::cout`

File: src/wesnoth.cpp (hunk 5 above)
Comment (babaissarkar) [reply src/wesnoth.cpp]: reverted.

File: src/wesnoth.cpp (hunk 7 above)
Comment (babaissarkar) [reply src/wesnoth.cpp]: Moved.

File: src/wesnoth.cpp (hunk 6 above)
Comment (babaissarkar) [reply src/wesnoth.cpp:1009]: deleted the `FONTCONFIG_PATH` setenv call since it was no longer needed.

File: src/desktop/battery_info.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vultraz) [reply src/desktop/battery_info.cpp:31]: I'm still seeing `SDL.h` for this one (in battery_info).

File: src/sdl/window.cpp [hunk 10]
```diff
@@ -97,48 +97,71 @@ window::~window()
...
-	SDL_Point res;
-	SDL_GetRendererOutputSize(*this, &res.x, &res.y);
+#ifdef __ANDROID__
+	return this->get_size();
+#else
+ 	SDL_Point res;
+ 	SDL_GetRendererOutputSize(*this, &res.x, &res.y);
 
-	return res;
+ 	return res;
```
[the remark this answers is not in the document]
Comment (Vultraz) [reply src/sdl/window.cpp:128]: Actually, looks like leading whitespace.

File: src/desktop/battery_info.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/desktop/battery_info.cpp]: Fixed.

File: src/wesnoth.cpp (hunk 6 above)
Comment (soliton-) [reply src/wesnoth.cpp:1008]: If it's hard to set env variables then if someone goes through the trouble it would be nice to respect their wish? Why would you expect random env variable values?

Comment (soliton-) [reply src/wesnoth.cpp:1008]: If you truely expect random env variable values then I suggest to log those so any issue resulting from them can be traced. There are plenty more env variables that can change how wesnoth runs.

File: src/filesystem.cpp [hunk 11]
```diff
@@ -668,6 +672,11 @@ static void setup_user_data_dir()
 #if defined(__APPLE__) && !defined(__IPHONEOS__)
 	migrate_apple_config_directory_for_unsandboxed_builds();
 #endif
+
+#ifdef __ANDROID__
+	user_data_dir = bfs::path(SDL_AndroidGetExternalStoragePath());
+#endif
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/filesystem.cpp]: Moved.

File: src/filesystem.cpp [hunk 12]
```diff
@@ -822,6 +831,11 @@ void set_cache_dir(const std::string& newcachedir)
 
 static const bfs::path& get_user_data_path()
 {
+#ifdef __ANDROID__
+	if (user_data_dir.empty()) {
+		user_data_dir = bfs::path(SDL_AndroidGetExternalStoragePath());
+	}
+#endif
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/filesystem.cpp]: Fixed by setting userdata dir in `set_user_data_dir()`, so this has been removed.

File: src/sdl/window.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/sdl/window.cpp]: Fixed.

File: src/wesnoth.cpp (hunk 6 above)
Comment (babaissarkar) [reply src/wesnoth.cpp]: Updated to not overwrite Env Vars. (1 -> 0)

**Comment by soliton-** [comment]
Btw, is the expectation a squash merge or should the individual commits remain? Even if it'd be a huge commit the former would be fine IMO if a good commit message is written. If the latter then some consolidation of commits would be needed.

**Comment by babaissarkar** [comment]
> Btw, is the expectation a squash merge or should the individual commits remain? Even if it'd be a huge commit the former would be fine IMO if a good commit message is written. If the latter then some consolidation of commits would be needed.

I think the latter would make the commits more understandable, that is a small number of medium sized commits. I'll try to consolidate my commits at least. Feel free to suggest which one's sorely need squashing or any other relevant ideas.

@loonycyborg thoughts? I'd prefer if you'd squish your own commits.

**Comment by loonycyborg** [comment]
I generally prefer to avoid squashing commits for the purpose of more precise git blame.

**Comment by soliton-** [comment]
To be clear I'm all for not squashing everything but at a minimum fixup commits should be combined with the commit they're fixing before merge then.

**Comment by soliton-** [comment]
Sure

**Comment by babaissarkar** [comment]
## Task List 2
### InitActivity
- [x] ~~Redownload menu implementation~~ `Tap to Start` option redownloads anyway.
- [x] Crash after first download/install. Second time launches fine.
- [x] Key for manual download in `status.properties` and install zip and run correctly without a statusfile in this case.
- [x] Help message when clicking "Install from ZIP".

**Review by Pentarctagon — APPROVED** [review]

File: packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java [hunk 13]
```diff
@@ -0,0 +1,525 @@
...
+	private int progress = 0;
+	private long max = 0;
+
+	private final static LinkedHashMap<String, String> packages = new LinkedHashMap<String, String>();
+	private final String archiveURL =
+		"https://sourceforge.net/projects/wesnoth/files/android/%s/download";
+
+	private File dataDir;
+
+	static {
```
[the remark this answers is not in the document]
Comment (Pentarctagon) [reply packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java:72]: Does it need to be a LinkedHashMap specifically? Otherwise there's `Map.of()`.

[the remark this answers is not in the document]
Comment (babaissarkar) [reply packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java:72]: > Does it need to be a LinkedHashMap specifically? Otherwise there's `Map.of()`.

Android is stuck with Java 8. For 9+ feature we'd need [desugaring](https://developer.android.com/studio/write/java8-support-table).

> dont like such static block, but if we have it then better together with field

Sure. Alternatively I can move it inside the `onCreate()` method, no `static`.

File: packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java [hunk 14]
```diff
@@ -0,0 +1,525 @@
...
+					certFile.createNewFile();
+					copyStream(
+						getResources().openRawResource(R.raw.cacert),
+						new FileOutputStream(certFile));
+				}
+				
+				// Store status
+				status.store(new FileOutputStream(statusFile), "Wesnoth Assets Status");
+			} catch (Exception e) {
+				Log.e("InitActivity", "Exception", e);
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java:258]: > what is the point of logging exceptions? Where user should look for them

They are stored in Android logcat. (`adb logcat`)

File: packaging/android/app/src/main/java/org/wesnoth/Wesnoth/WesnothActivity.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+// Needs to be inside an activity so we can use `startActivity`.
+public void open(String url) {
+	Log.d("WesnothActivity", "opening " + url);
+	Intent openIntent = new Intent(Intent.ACTION_VIEW);
+	openIntent.setData(Uri.parse(url));
+	startActivity(openIntent);
+}
+
+public double getBatteryPercentage() {
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply packaging/android/app/src/main/java/org/wesnoth/Wesnoth/WesnothActivity.java:38]: Used via JNI, see https://github.com/wesnoth/wesnoth/blob/6e4eb85bc04a2a7efe4c0aa413bf41e5f8f42841/src/desktop/battery_info.cpp#L61-L68.

File: packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java [hunk 16]
```diff
@@ -0,0 +1,525 @@
...
+import android.widget.TextView;
+import android.widget.Toast;
+
+public class InitActivity extends Activity {
+	private int length = 0;
+	private int progress = 0;
+	private long max = 0;
+
+	private final static LinkedHashMap<String, String> packages = new LinkedHashMap<String, String>();
+	private final String archiveURL =
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java]: done

File: packaging/android/app/src/main/java/org/wesnoth/Wesnoth/WesnothActivity.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+import android.os.Build;
+import android.util.Log;
+
+import org.libsdl.app.SDLActivity;
+
+public class WesnothActivity extends SDLActivity
+{
+
+// Needs to be inside an activity so we can use `startActivity`.
+public void open(String url) {
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply packaging/android/app/src/main/java/org/wesnoth/Wesnoth/WesnothActivity.java]: indented.

File: packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java [hunk 18]
```diff
@@ -0,0 +1,525 @@
...
+			// Extract certificates file
+			try {
+				File certDir = new File(dataDir, "certificates");
+				if (!certDir.exists()) {
+					certDir.mkdir();
+				}
+				File certFile = new File(certDir, "cacert.pem");
+				if (!certFile.exists()) {
+					certFile.createNewFile();
+					copyStream(
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java]: closed stream via try-with-resources.

File: packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java [hunk 19]
```diff
@@ -0,0 +1,525 @@
...
+		progressText.setText("Connecting...");
+		
+		Executors.newSingleThreadExecutor().execute(() -> {
+			//TODO Update mechanism when patch is available.
+
+			Properties status = new Properties();
+			File statusFile = new File(dataDir, "status.properties");
+			try {
+				if (statusFile.exists()) {
+					status.load(new FileInputStream(statusFile));
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java]: closed stream via try-with-resources.

File: src/scripting/plugins/manager.cpp [hunk 20]
```diff
@@ -45,7 +45,9 @@ plugins_manager::plugins_manager(application_lua_kernel * kernel)
 	, playing_()
 	, kernel_(kernel)
 {
+#ifndef __ANDROID__
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/scripting/plugins/manager.cpp:48]: Seems like it's no longer needed. Will be removed.

File: packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java [hunk 21]
```diff
@@ -0,0 +1,525 @@
...
+import android.view.animation.AnimationUtils;
+import android.widget.Button;
+import android.widget.ImageButton;
+import android.widget.PopupMenu;
+import android.widget.ProgressBar;
+import android.widget.TextView;
+import android.widget.Toast;
+
+public class InitActivity extends Activity {
+	private int length = 0;
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java]: Making these local causes `error: local variables referenced from a lambda expression must be final or effectively final` error.

[the remark this answers is not in the document]
Comment (ProditorMagnus) [reply packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java]: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicInteger.html

[the remark this answers is not in the document]
Comment (babaissarkar) [reply packaging/android/app/src/main/java/org/wesnoth/Wesnoth/InitActivity.java]: Removed global vars: https://github.com/wesnoth/wesnoth/pull/7707/commits/a5ec0cbd6cf6d5ad0a4bdef26a6417ecd6c25cdb.

**Comment by babaissarkar** [comment]
Manually merged to `master`.

**Comment by Zireael07** [comment]
If it was merged, why does GH show it as closed?

**Comment by babaissarkar** [comment]
> If it was merged, why does GH show it as closed?

Github issue: for some weird reason the rebase merge button got disabled, despite my attempts at fixing it. So I manually merged the commits from here into master and force pushed master into this branch, which flagged it as closed, despite commits existing on master.

TL;DR: github limitation

**Comment by soliton-** [comment]
The normal thing to do is rebase the PR on master and force push to the PR. Then you can fast forward merge and push to master exactly what's in the PR so github will close it.
